### PR TITLE
KubeEdge based Edgecluster POC

### PR DESCRIPTION
**Pull request by chenqianfzh**
This PR serves as a PoC to show a kubeedge node can be extended as an edge cluster.

### Verfication Steps:
1. Create a cloud node and an edge node per instructions at https://github.com/pdgetrf/ArktosEdge/wiki/KubeEdge-Setup-(Same-Network).
2. in the cloud node, create the crd (it is assumed that the working directory is the kubeedge repo root):
```
kubectl apply -f build/crds/edgecluster/mission_v1.yaml
```
3. In the edge node, update the file /etc/kubeedge/config/edgecore.yaml following the example at tests/edgecluster/data/sample_edgecore_with_edgecluster.yaml, to add the sub-section of edgeCluster under the section of edged.
4. start cloudcore in the cloud node
5. start edgecore in the edge node
6. apply the mission yamls under tests/edgecluster/data/missions. It was verifed that the missions are deployed to the target edge cluster as expected. Namely

        a. the mission content is running in the edge cluster when the mission is applied in the host cluster.
        b. the mission content is gone in the edge cluster when the mission is deleted in the host cluster.
        c. the mission content will be run in the edge cluster only if the edge cluster meets the placement criteria in the mission yaml.

### CODE REVIEW

File: pkg/apis/componentconfig/edgecore/v1alpha1/types.go [hunk 1]
```diff
@@ -471,3 +474,22 @@ type EdgeStream struct {
...
+type EdgeCluster struct {
+	// Enable indicates whether edge cluster is enabled, if set to false, skip checking other configs.
+	// default true
+	Enable bool `json:"enable"`
+
+	// clusterKubeconfig indicates the path to the edge cluster kubeconfig file
+	ClusterKubeconfig string `json:"clusterKubeconfig,omitempty"`
+
+	// Distribution of the cluster, supported value: arkots, to support in the furture: k3s, 
+	KubeDistribution string `json:"kubeDistribution,omitempty"`
```
Comment (pdgetrf) [inline pkg/apis/componentconfig/edgecore/v1alpha1/types.go:488]: nit: Distribution --> Distro

File: pkg/apis/componentconfig/edgecore/v1alpha1/types.go [hunk 2]
```diff
@@ -471,3 +474,22 @@ type EdgeStream struct {
...
+
+	// Name of the edge cluster
+	ClusterName string `json:"clusterName,omitempty"`
```
Comment (pdgetrf) [inline pkg/apis/componentconfig/edgecore/v1alpha1/types.go:491]: Consider remove "Cluster" in the name. same with ClusterLabels as both are already in EdgeCluster struct

File: cloud/pkg/edgecontroller/controller/downstream.go [hunk 3]
```diff
@@ -545,6 +548,61 @@ func (dc *DownstreamController) syncRuleEndpoint() {
...
+				if err != nil {
+					klog.Warningf("Built message resource failed with error: %v", err)
+					return true
+				}
+				msg.BuildRouter(modules.EdgeControllerModuleName, constants.GroupResource, resource, operation)
+				msg.Content = mission
+				if err := dc.messageLayer.Send(*msg); err != nil {
+					klog.Warningf("Send message failed with error: %s, operation: %s, resource: %s", err, msg.GetOperation(), msg.GetResource())
+				} else {
+					klog.V(4).Infof("Send message successfully, operation: %s, resource: %s", msg.GetOperation(), msg.GetResource())
```
Comment (pdgetrf) [inline cloud/pkg/edgecontroller/controller/downstream.go:598]: "message sent successfully"

also I'm a bit confused about this whole section. what does "all nodes" mean here?

**Review by pdgetrf — APPROVED** [review]
lgtm

a few cosmetic suggestion.

File: pkg/apis/componentconfig/edgecore/v1alpha1/types.go (hunk 1 above)
Comment (chenqianfzh) [reply pkg/apis/componentconfig/edgecore/v1alpha1/types.go]: updated.

File: pkg/apis/componentconfig/edgecore/v1alpha1/types.go (hunk 2 above)
Comment (chenqianfzh) [reply pkg/apis/componentconfig/edgecore/v1alpha1/types.go]: Yep, I made the change.

File: cloud/pkg/edgecontroller/controller/downstream.go (hunk 3 above)
Comment (chenqianfzh) [reply cloud/pkg/edgecontroller/controller/downstream.go]: Updated all the "Send message successfully" in the code. 

"all the nodes" means all the edge nodes in the cluster. One edge controller may talk to multiple edge nodes.

**Pull request merged**